opnode/l2/api.go

```diff
@@ -0,0 +1,297 @@
+package l2
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/ethereum-optimism/optimistic-specs/opnode/eth"
+
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/holiman/uint256"
+
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+type ErrorCode int
+
+const (
+	UnavailablePayload ErrorCode = -32001
+)
+
+type Bytes32 [32]byte
+
+func (b *Bytes32) UnmarshalJSON(text []byte) error {
+	return hexutil.UnmarshalFixedJSON(reflect.TypeOf(b), text, b[:])
+}
+
+func (b *Bytes32) UnmarshalText(text []byte) error {
+	return hexutil.UnmarshalFixedText("Bytes32", text, b[:])
+}
+
+func (b Bytes32) MarshalText() ([]byte, error) {
+	return hexutil.Bytes(b[:]).MarshalText()
+}
+
+func (b Bytes32) String() string {
+	return hexutil.Encode(b[:])
+}
+
+type Bytes256 [256]byte
+
+func (b *Bytes256) UnmarshalJSON(text []byte) error {
+	return hexutil.UnmarshalFixedJSON(reflect.TypeOf(b), text, b[:])
+}
+
+func (b *Bytes256) UnmarshalText(text []byte) error {
+	return hexutil.UnmarshalFixedText("Bytes32", text, b[:])
+}
+
+func (b Bytes256) MarshalText() ([]byte, error) {
+	return hexutil.Bytes(b[:]).MarshalText()
+}
+
+func (b Bytes256) String() string {
+	return hexutil.Encode(b[:])
+}
+
+type Uint64Quantity = hexutil.Uint64
+
+type BytesMax32 []byte
+
+func (b *BytesMax32) UnmarshalJSON(text []byte) error {
+	if len(text) > 64+2+2 { // account for delimiter "", and 0x prefix
+		return fmt.Errorf("input too long, expected at most 32 hex-encoded, 0x-prefixed, bytes: %x", text)
+	}
+	return (*hexutil.Bytes)(b).UnmarshalJSON(text)
+}
+
+func (b *BytesMax32) UnmarshalText(text []byte) error {
+	if len(text) > 64+2 { // account for 0x prefix
+		return fmt.Errorf("input too long, expected at most 32 hex-encoded, 0x-prefixed, bytes: %x", text)
+	}
+	return (*hexutil.Bytes)(b).UnmarshalText(text)
+}
+
+func (b BytesMax32) MarshalText() ([]byte, error) {
+	return (hexutil.Bytes)(b).MarshalText()
+}
+
+func (b BytesMax32) String() string {
+	return hexutil.Encode(b)
+}
+
+type Uint256Quantity = uint256.Int
+
+type Data = hexutil.Bytes
+
+// TODO: implement neat 8 byte typed payload ID and upstream it to geth api definitions
```

Comment:
Is this is a significant effort, could we do it now?

Reply:
This is part of what I already did in the PR to upstream geth (see PR status update in slack). Can merge this as-is and upgrade later.